### PR TITLE
win_stat: Rewrite using AnsibleModule

### PR DESCRIPTION
**Pull request by dagwieers**
##### SUMMARY
Simple rewrite using AnsibleModule.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
win_stat

### CODE REVIEW

File: lib/ansible/modules/windows/win_stat.ps1 [hunk 1]
```diff
@@ -13,28 +13,33 @@ function DateTo-Timestamp($start_date, $end_date) {
...
+$path = $module.Params.path
+$get_md5 = $module.Params.get_md5
+$get_checksum = $module.Params.checksum
+$checksum_algortihm = $module.Params.checksum_algorithm
+
+$module.Result.stat = @{ exists=$false }
+
 # get_md5 will be an undocumented option in 2.9 to be removed at a later
 # date if possible (3.0+)
 if (Get-Member -inputobject $params -name "get_md5") {
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_stat.ps1]: @jborean93 How do we do this in the AnsibleModule world ?

Comment (dagwieers) [reply lib/ansible/modules/windows/win_stat.ps1]: We probably need to expose `RawParams` in the `AnsibleModule` object.

File: lib/ansible/modules/windows/win_stat.ps1 [hunk 2]
```diff
@@ -3,7 +3,7 @@
 # Copyright: (c) 2017, Ansible Project
 # GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
 
-#Requires -Module Ansible.ModuleUtils.Legacy
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_stat.ps1:6]: This module uses the `Get-FileChecksum` https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/powershell/Ansible.ModuleUtils.Legacy.psm1#L317 cmdlet in Legacy.psm1. Best to keep this import here until we move that out into FileUtil.

File: lib/ansible/modules/windows/win_stat.ps1 [hunk 3]
```diff
@@ -13,28 +13,27 @@ function DateTo-Timestamp($start_date, $end_date) {
...
-$result = @{
-    changed = $false
-    stat = @{
-        exists = $false
+$spec = @{
+    options = @{
+        path = @{ type='path'; required=$true; aliases=@( 'dest', 'name' ) }
+        get_checksum = @{ type='bool'; default=$true }
+        checksum_algorithm = @{ type='str'; default='sha1'; choices=@( 'md5', 'sha1', 'sha256', 'sha384', 'sha512' ) }
+        get_md5 = @{ type='bool'; default=$false; removed_in_version='2.9' }
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_stat.ps1:41]: I don't think we can use `removed_in_version` for `get_md5`. The actual option is set to become "undocumented" in 2.9 but still available due to win_stat and stat being a stableinterface and people wanting a very slow removal.

File: lib/ansible/modules/windows/win_stat.ps1 [hunk 4]
```diff
@@ -105,14 +104,14 @@ If ($info -ne $null) {
             try {
                 $stat.md5 = Get-FileChecksum -path $path -algorithm "md5"
             } catch {
-                Fail-Json -obj $result -message "failed to get MD5 hash of file, remove get_md5 to ignore this error: $($_.Exception.Message)"
+                $module.FailJson("Failed to get MD5 hash of file, remove get_md5 to ignore this error: $($_.Exception.Message)")
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_stat.ps1]: Add `, $_)` as a 2nd argument to ensure the exception info is returned on higher verbosity levels.

File: lib/ansible/modules/windows/win_stat.ps1 [hunk 5]
```diff
@@ -105,14 +104,14 @@ If ($info -ne $null) {
...
             }
         }
         if ($get_checksum) {
             try {
                 $stat.checksum = Get-FileChecksum -path $path -algorithm $checksum_algorithm
             } catch {
-                Fail-Json -obj $result -message "failed to get hash of file, set get_checksum to False to ignore this error: $($_.Exception.Message)"
+                $module.FailJson("Failed to get hash of file, set get_checksum to False to ignore this error: $($_.Exception.Message)")
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_stat.ps1]: Same here, add `$_` as a 2nd argument to `FailJson`.

File: lib/ansible/modules/windows/win_stat.ps1 (hunk 3 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_stat.ps1:41]: Ok, makes sense.

File: lib/ansible/modules/windows/win_stat.ps1 (hunk 2 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_stat.ps1:6]: Fixed.

File: lib/ansible/modules/windows/win_stat.ps1 (hunk 3 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_stat.ps1:41]: The actual problem here is that we can no longer check if `get_md5` was being set. So if we want to do this in the new world order, we need a way to check if parameters were set by the user.

File: lib/ansible/modules/windows/win_stat.ps1 (hunk 4 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_stat.ps1]: Fixed.

File: lib/ansible/modules/windows/win_stat.ps1 (hunk 5 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_stat.ps1]: Fixed.

File: lib/ansible/modules/windows/win_stat.ps1 (hunk 3 above)
Comment (jborean93) [reply lib/ansible/modules/windows/win_stat.ps1:41]: You can still check it by not setting a default for `get_md5` and seeing if it is `$null`. If it is null we would just need to manually set to `$false` until we eventually remove it.

**Comment by mattclay** [comment]
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/99139/29/tests

**Comment by jborean93** [comment]
@dagwieers, I pushed a commit that rebased this PR and fixed the issue with CI. The line https://github.com/ansible/ansible/pull/49653/files#diff-c1c59bbe17c1b0da310272cdaa0b0bacR50 was trying to get the value from `$module.Params.checksum` instead of `$module.Params.get_checksum`. I also fixed a few other minor things that came up later in the tests. Thanks for working on this.

**Pull request merged**